ballerina/natives.bal

```diff
@@ -41,16 +72,30 @@ public isolated function matches(string stringToMatch, string regex) returns boo
 # + replacement - The replacement string to replace the substrings, which
 #                 match the regex
 # + return - The resultant string with the replaced substrings
-public isolated function replaceAll(string originalString, string regex, string replacement) returns string {
-    handle value = replaceAllExternal(java:fromString(originalString), java:fromString(regex),
-                                      java:fromString(replacement));
-    string? updatedString = java:toString(value);
-    if updatedString is string {
+public isolated function replaceAll(string originalString, string regex, Replacement replacement) returns string {
+    if (replacement is ReplacerFunction) {
+        string updatedString = "";
+        int startIndex = 0;
+        Match[] matchedArray = searchAll(originalString, regex);
+        foreach Match matched in matchedArray {
+            updatedString += strings:substring(originalString, startIndex, matched.startIndex) +
+                                               getStringFromReplacerFunction(matched, replacement);
+            startIndex = matched.endIndex;
+        }
+        if (startIndex < originalString.length()) {
+            updatedString += strings:substring(originalString, startIndex, originalString.length());
+        }
         return updatedString;
     } else {
-        // Should never reach here.
-        error e = error(string `error occurred while replacing ${regex} in ${originalString}`);
-        panic e;
+        handle|error value = trap replaceAllExternal(java:fromString(originalString), java:fromString(regex),
+                                                     java:fromString(replacement));
```

Comment:
Can't we use the same logic in both function and string values?

Reply:
Without checking the type of the replacement, Can't we just pass the replacement to the `getNewSubstring` method and get the string value? I changed the function to return the replacement string.

```
        string updatedString = "";
        int startIndex = 0;
        Match[] matchedArray = searchAll(originalString, regex);
        foreach Match matched in matchedArray {
            updatedString += strings:substring(originalString, startIndex, matched.startIndex) +
                                               getNewSubstring(matched, replacement);
            startIndex = matched.endIndex;
        }
        if (startIndex < originalString.length()) {
            updatedString += strings:substring(originalString, startIndex, originalString.length());
        }
        return updatedString;
```

Reply:
Updated now.